### PR TITLE
fix: validator signature

### PR DESCRIPTION
**Pull request by christopherbrumm**

### CODE REVIEW

File: x/core/01_interchain_security/types/multisig.go [hunk 1]
```diff
@@ -31,6 +31,8 @@ func VerifyMultisig(validators []string, threshold uint32, signatures [][]byte,
 		validatorAddresses[strings.ToLower(address)] = true
 	}
 
+	signedValidators := make(map[string]bool)
```
Comment (yorhodes) [inline x/core/01_interchain_security/types/multisig.go:34]: can we remove `validatorAddresses` in favor of this? also why use this approach as opposed to the 2 pointer iterator approach in the solidity implementation which works with the array type

File: x/core/keeper/logic_message.go [hunk 2]
```diff
@@ -22,6 +22,11 @@ func (k Keeper) ProcessMessage(ctx sdk.Context, mailboxId util.HexAddress, rawMe
 	if err != nil {
 		return fmt.Errorf("failed to find mailbox with id: %s", mailboxId.String())
 	}
+
+	if message.Destination != mailbox.LocalDomain {
+		return fmt.Errorf("message destination %v does not match local domain %v", message.Destination, mailbox.LocalDomain)
+	}
```
Comment (yorhodes) [inline x/core/keeper/logic_message.go:28]: in the spirit of atomic PRs I think this should be separated into its own fix

File: x/core/01_interchain_security/types/multisig.go [hunk 3]
```diff
@@ -39,8 +41,11 @@ func VerifyMultisig(validators []string, threshold uint32, signatures [][]byte,
 
 		addressBytes := crypto.PubkeyToAddress(*recoveredPubkey)
 		address := util.EncodeEthHex(addressBytes[:])
-		if validatorAddresses[address] {
+
+		// Ensure that the validator has not signed already
+		if validatorAddresses[address] && !signedValidators[address] {
 			validSignatures++
+			signedValidators[address] = true
```
Comment (yorhodes) [inline x/core/01_interchain_security/types/multisig.go:48]: you should add a test that attempts to pass duplicate signatures
see https://github.com/hyperlane-xyz/v3-docs/pull/356

File: x/core/keeper/logic_message.go (hunk 2 above)
Comment (yorhodes) [reply x/core/keeper/logic_message.go:28]: this should also be accompanied by a test

**Review by yorhodes — CHANGES_REQUESTED** [review]

File: x/core/keeper/logic_message.go (hunk 2 above)
Comment (christopherbrumm) [reply x/core/keeper/logic_message.go:28]: Agree!

File: x/core/01_interchain_security/types/multisig.go (hunk 3 above)
Comment (christopherbrumm) [reply x/core/01_interchain_security/types/multisig.go:48]: Agree!

File: x/core/keeper/logic_message.go (hunk 2 above)
Comment (christopherbrumm) [reply x/core/keeper/logic_message.go]: Fixed in #69

File: x/core/01_interchain_security/types/merkle_root_multisig.go [hunk 4]
```diff
@@ -130,6 +136,33 @@ func (m *MerkleRootMultisigMetadata) digest(message *util.HyperlaneMessage) [32]
...
+
+	signedIndex := make([]byte, 4)
+	binary.BigEndian.PutUint32(signedIndex, m.SignedIndex)
+
+	var signaturesBytes []byte
+	for _, sig := range m.Signatures {
+		signaturesBytes = append(signaturesBytes, sig...)
+	}
+
+	return slices.Concat(
```
Comment (mbreithecker) [inline x/core/01_interchain_security/types/merkle_root_multisig.go:156]: Maybe it makes sense to return a full copy here

File: x/core/01_interchain_security/types/merkle_root_multisig_test.go [hunk 5]
```diff
@@ -116,8 +168,90 @@ var _ = Describe("msg_server.go", Ordered, func() {
...
+		// Act
+		verify, err := merkleRootMultiSig.Verify(sdk.Context{}, metadata.Bytes(), message)
+
 		// Assert
 		Expect(err).To(BeNil())
 		Expect(verify).To(BeFalse())
 	})
 })
+
+func signDigest(digest []byte, privateKeyHex string) []byte {
```
Comment (mbreithecker) [inline x/core/01_interchain_security/types/merkle_root_multisig_test.go:226]: Maybe it makes sense to put test utils into types_suite_test.go as it might also be used by the message_id_tests

File: x/core/01_interchain_security/types/multisig.go [hunk 6]
```diff
@@ -14,40 +15,32 @@ type MultisigISM interface {
...
 		}
 
-		addressBytes := crypto.PubkeyToAddress(*recoveredPubkey)
-		address := util.EncodeEthHex(addressBytes[:])
-		if validatorAddresses[address] {
-			validSignatures++
+		signerBytes := crypto.PubkeyToAddress(*recoveredPubkey)
+		signer := util.EncodeEthHex(signerBytes[:])
+
+		for validatorIndex < validatorCount && signer != strings.ToLower(validators[validatorIndex]) {
```
Comment (mbreithecker) [inline x/core/01_interchain_security/types/multisig.go:34]: Please add some comments

File: x/core/01_interchain_security/types/multisig.go [hunk 7]
```diff
@@ -14,40 +15,32 @@ type MultisigISM interface {
 }
 
 func VerifyMultisig(validators []string, threshold uint32, signatures [][]byte, digest [32]byte) (bool, error) {
```
Comment (mbreithecker) [inline x/core/01_interchain_security/types/multisig.go:17]: Add comment here

File: x/core/01_interchain_security/types/multisig.go (hunk 1 above)
Comment (christopherbrumm) [reply x/core/01_interchain_security/types/multisig.go]: Used 2 pointer iterator to be fully aligned with the solidity implementation.

File: x/core/01_interchain_security/types/merkle_root_multisig.go (hunk 4 above)
Comment (christopherbrumm) [reply x/core/01_interchain_security/types/merkle_root_multisig.go:156]: That approach already returns a full copy, as `slices.Concat` returns a new slice without referencing any of the original slices directly.

**Review by mbreithecker — APPROVED** [review]
